flows/src/main/kotlin/nl/nuts/consent/flow/ConsentFlows.kt

```diff
@@ -55,24 +57,48 @@ object ConsentFlows {
         return state.referencedAttachments(state.attachments.map { it to serviceHub.attachments.openAttachment(it)!! }.toMap())
     }
 
-    private fun containedConsentRecordHash(state : ConsentBase, serviceHub: ServiceHub) : Set<String> {
-        return state.attachments.map {
+    private fun containedConsentRecordHash(attachments: Set<SecureHash>, serviceHub: ServiceHub) : Set<String> {
+        val hashes = mutableSetOf<String>()
+
+        attachments.forEach {
             val att = serviceHub.attachments.openAttachment(it)!!
-            val metadata = extractMetadata(att)
-            metadata.consentRecordHash
-        }.toSet()
+            if (att !is ContractAttachment) {
+                val metadata = extractMetadata(att)
+                hashes.add(metadata.consentRecordHash)
+            }
+        }
+
+        return hashes
     }
 
     /**
      * A check for duplicate attachments, this is done outside the contract because the contents of the attachments is required
+     * It recursively checks previous transactions as well
      */
-    fun raiseOnDuplicateConsentRecords(consentState: ConsentState, consentBranch: ConsentBranch, serviceHub: ServiceHub, side:String = "origin") {
-        val consentRecordHashInState = containedConsentRecordHash(consentState, serviceHub)
-        val consentRecordHashInBranch = containedConsentRecordHash(consentBranch, serviceHub)
+    fun raiseOnDuplicateConsentRecords(stateRef: StateRef, attachments: Set<SecureHash>, serviceHub: ServiceHub, side:String = "origin") {
+        // new hashes
+        val newHashes = containedConsentRecordHash(attachments, serviceHub)
```

Comment:
How expensive is this? Can it be a parameter for this method so it runs only once?

Reply:
it has to fetch transaction -> states -> transaction -> etc, each being a db call.